src/site/content/en/blog/timer-throttling/index.md

```diff
@@ -0,0 +1,176 @@
+---
+layout: post
+title: Timer throttling
+authors:
+  - jakearchibald
+description: >
+  Chrome 88 will heavily throttle chained JavaScript timers for hidden pages.
+date: 2020-12-03
+updated: 2020-12-03
+hero: hero.jpg
+alt: A collection of broken clocks
+tags:
+  - blog
+  - javascript
+  - timers
+feedback:
+  - api
+---
+
+Chrome 88 (January 2020) will heavily throttle chained JavaScript timers for hidden pages in particular conditions. This will reduce CPU usage, which will also reduce battery usage. There are some edge cases where this will change behavior, but timers are often used where a different API would be more efficient, and more reliable.
+
+Ok, that was pretty jargon heavy, and a bit ambiguous. Let's dig in…
+
+## Terminology
+
+### Hidden pages
+
+Generally, 'hidden' means a different tab is active, or the window has been minimized, but browsers may consider a page hidden whenever its content is totally not-visible. Some browsers go further than others here, but you can always use the [page visibility API](https://developer.mozilla.org/en-US/docs/Web/API/Page_Visibility_API) to track when the browser thinks visibility has changed.
+
+### JavaScript timers
+
+By "JavaScript timers" I mean `setTimeout` and `setInterval`, which allow you to schedule a callback sometime in the future. Timers are useful, and they aren't going away, but sometimes they're used to poll state when an event would be more efficient, and more accurate.
+
+### Chained timers
+
+If you call `setTimeout` in the same task as a `setTimeout` callback, the second invocation is 'chained'. With `setInterval`, each iteration is part of the chain. This might be easier to understand with code:
+
+```js
+let chainCount = 0;
+
+setInterval(() => {
+  console.log(`This is number ${chainCount} in the chain`);
+  chainCount++;
```

Comment:
To make the next parts of the article accurate, chainCount+ should be moved before console.log(), or should be initialized to 1 instead of 0.

If we execute this code on a page that has been hidden for 5 minutes:

```
let chainCount = 0;
let start = performance.now();

setInterval(() => {
  let t = Math.round(performance.now() - start);
  console.log(`t=${t} This is number ${chainCount} in the chain`);
  chainCount++;
}, 500);
```
We get an output that looks like this:

t=1216 This is number 0 in the chain
t=2215 This is number 1 in the chain
t=3216 This is number 2 in the chain
t=4215 This is number 3 in the chain
t=34216 This is number 4 in the chain   <--- intensive throttling was applied
t=48215 This is number 5 in the chain

If chainCount is incremented after console.log(), we get an output that matches the next parts of the article:

t=876 This is number 1 in the chain
t=1876 This is number 2 in the chain
t=2876 This is number 3 in the chain
t=3876 This is number 4 in the chain
t=31876 This is number 5 in the chain  <--- intensive throttling was applied
t=57876 This is number 6 in the chain

Sadly, the definition of "nesting level" in Chrome, Safari (and now Edge) doesn't match the spec https://crbug.com/1108877

Reply:
Fixed!